perses/dispersed/smc.py

```diff
@@ -0,0 +1,924 @@
+import simtk.openmm as openmm
+import openmmtools.cache as cache
+from typing import List, Tuple, Union, NamedTuple
+import os
+import copy
+import openmmtools.cache as cache
+from perses.dispersed.utils import *
+
+import openmmtools.mcmc as mcmc
+import openmmtools.integrators as integrators
+import openmmtools.states as states
+from openmmtools.states import ThermodynamicState, CompoundThermodynamicState, SamplerState
+import numpy as np
+import mdtraj as md
+from perses.annihilation.relative import HybridTopologyFactory
+import mdtraj.utils as mdtrajutils
+import pickle
+import simtk.unit as unit
+import tqdm
+from perses.tests.utils import compute_potential_components
+from openmmtools.constants import kB
+import pdb
+import logging
+import tqdm
+from sys import getsizeof
+import time
+from collections import namedtuple
+from perses.annihilation.lambda_protocol import LambdaProtocol
+from perses.annihilation.lambda_protocol import RelativeAlchemicalState, LambdaProtocol
+from perses.dispersed import *
+import random
+import pymbar
+import dask.distributed as distributed
+from perses.dispersed.parallel import Parallelism
+import tqdm
+import time
+# Instantiate logger
+logging.basicConfig(level = logging.NOTSET)
+_logger = logging.getLogger("sMC")
+_logger.setLevel(logging.DEBUG)
+
+#cache.global_context_cache.platform = openmm.Platform.getPlatformByName('Reference') #this is just a local version
+EquilibriumFEPTask = namedtuple('EquilibriumInput', ['sampler_state', 'inputs', 'outputs'])
+
+class SequentialMonteCarlo():
+    """
+    This class represents an sMC particle that runs a nonequilibrium switching protocol.
+    It is a batteries-included engine for conducting sequential Monte Carlo sampling.
+
+    WARNING: take care in writing trajectory file as saving positions to memory is costly.  Either do not write the configuration or save sparse positions.
+    """
+
+    supported_resampling_methods = {'multinomial': multinomial_resample}
+    supported_observables = {'ESS': ESS, 'CESS': CESS}
```

Comment:
could this be replaced in other parts of the code using `getattr()` rather than a dict of string onto object?

Reply:
i think this might be the better organizational strategy since I intend on placing several more resampling methods and observables into this class; also, i want to explicitly expose the observable and resampling methods available with user-input strings